apps/heft/src/cli/HeftActionRunner.ts

```diff
@@ -346,20 +370,50 @@ export class HeftActionRunner {
   }
 
   private async _executeOnceAsync(
-    executionManager: OperationExecutionManager,
+    executionManager: OperationExecutionManager<IHeftTaskOperationMetadata, IHeftPhaseOperationMetadata>,
     abortSignal: AbortSignal,
     requestRun?: (requestor?: string) => void
   ): Promise<OperationStatus> {
+    const { taskStart, taskFinish, phaseStart, phaseFinish } = this._internalHeftSession.lifecycle.hooks;
     // Record this as the start of task execution.
     this._metricsCollector.setStartTime();
     // Execute the action operations
     return await runWithLoggingAsync(
       () => {
-        const operationExecutionManagerOptions: IOperationExecutionOptions = {
+        const operationExecutionManagerOptions: IOperationExecutionOptions<
+          IHeftTaskOperationMetadata,
+          IHeftPhaseOperationMetadata
+        > = {
           terminal: this._terminal,
           parallelism: this._parallelism,
           abortSignal,
-          requestRun
+          requestRun,
+          beforeExecuteOperationAsync: async (operation: Operation<IHeftTaskOperationMetadata>) => {
+            if (operation.metadata.task && taskStart.isUsed()) {
+              await taskStart.promise({ operation });
+            }
+          },
+          afterExecuteOperationAsync: async (operation: Operation<IHeftTaskOperationMetadata>) => {
+            if (operation.metadata.task && taskFinish.isUsed()) {
+              await taskFinish.promise({ operation });
+            }
+          },
+          beforeExecuteOperationGroupAsync: async (
+            operationGroup: OperationGroupRecord<IHeftPhaseOperationMetadata>,
+            operation: Operation<IHeftPhaseOperationMetadata>
+          ) => {
+            if (operation.metadata.phase && phaseStart.isUsed()) {
```

Comment:
```suggestion
            if (operationGroup.metadata.phase && phaseStart.isUsed()) {
```

Reply:
Do these group hooks even need to receive the phase marker operation at all? I guess it does include the clean step.